bayes_opt/bayesian_optimization.py

```diff
@@ -260,21 +283,146 @@ def maximize(self,
+        if self._dataset is not None:
+            exact_x_dict = []
+            while not self._queue.empty or iteration < n_iter:
+                try:
+                    x_probe = next(self._queue)
+                except StopIteration:
+                    util.update_params()
+                    x_probe = self.suggest(util)
+                    iteration += 1
+
+                try:
+                    exact_x_dict.append(dict(zip(self._space.keys, x_probe.T)))
+
+                except AttributeError:
+                    exact_x_dict.append(x_probe)
+
+                approximation = self.get_approximation(self._dataset, x_probe)
+                if self._target_column is not None and approximation is not None:
+
+                    self._space.register(approximation[0], approximation[1])
+
+                elif approximation is not None:
+                    self.probe(approximation, lazy=False)
+                else:
+                    self.probe(x_probe, lazy=False)
+
+                if self._bounds_transformer:
+                    self.set_bounds(
+                        self._bounds_transformer.transform(self._space))
+            self.dispatch(Events.OPTIMIZATION_END)
+            self.save_res_to_csv(True, exact_x=exact_x_dict)
+
+        else:
+            while not self._queue.empty or iteration < n_iter:
+                try:
+                    x_probe = next(self._queue)
+                except StopIteration:
+                    util.update_params()
+                    x_probe = self.suggest(util)
+                    iteration += 1
+
+                self.probe(x_probe, lazy=False)
+
+                if self._bounds_transformer:
+                    self.set_bounds(
+                        self._bounds_transformer.transform(self._space))
+
+            self.dispatch(Events.OPTIMIZATION_END)
+            self.save_res_to_csv(False)
+
+    def get_approximation(self, dataset, x_probe):
+        """
+        Method to get from the dataset passed by the user the nearest point to the x_probe point
+
+        Parameters
+        ----------
+
+        dataset: pandas.DataFrame
+            dataset specified by the user
+
+        x_probe: dict
+            point found by the optimization process
+        """
+
+        try:
+            x_array = numpy.array(list(x_probe.values()))
+
+        except AttributeError:
+            x_array = x_probe
+
+        min_distance = None
+        min_index = None
+        approximations = []
+
+        if self._target_column is None:
+            for row in dataset.itertuples():
+                dataset_tuple = []
+                for j in range(x_array.size):
+                    dataset_tuple.append(row[j + 1])
+
+                res = numpy.linalg.norm(x_array - dataset_tuple, 2)
+
+                if min_distance is None:
+                    min_distance = res
+                    approximations = self._space.array_to_params(dataset_tuple)
```

Comment:
I feel this should be `approximations = [ self._space... ]`, i.e. with the enclosing `[ ]` on the right-hand side, since you declared `approximations` as a list, not a dict. Can you even append values to a dict? Does `random.choice()` work with a dict? You can also change the declaration of `approximations` to an empty dict if this currently works

Reply:
notation fixed for the assignment statement of approximations